keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,548 @@
+
+A feature can be registered as server level feature or cluster level feature, but not both.
+For data compatibility, we do not support changing a feature gate from one type to another. 
+If a developer finds out that their feature needs to be a cluster level feature instead of server level after a release, they should add a new feature and use a new feature name.
+
+```go
+DefaultEtcdServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+DefaultEtcdClusterFeatureGates := map[Feature]VersionedSpecs {
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+### Set the Feature Gates
+To set the correct user expectation, we will add a new flag `--cluster-feature-gates` (+`config-file` configuration support) similar to the server `--feature-gates` to set the values of cluster feature gates as expected by the local server. The server will use this flag value to propose `proposed_feature_gates` to the cluster. The final values of the cluster feature gate are actually decided through raft consensus.
+
+We do not support dynamically changing the feature gates when the server is running.
+
+### Consensus Algorithm
+
+To guarantee consistent value of if a feature is enabled in the whole cluster, the leader would decide if the feature is enabled for all cluster members, and propagate the information to all members. 
+
+1. When an etcd server starts, the cluster feature gate would be set to nil, which means all cluster level features are disabled.
+
+1. When there is an update in `ClusterVersion` or the server itself, the server would propose the values of cluster feature gates by publishing a new `Attributes.feature_gates` field [through raft](https://github.com/etcd-io/etcd/blob/e37a67e40b3f5ff8ef81f9de6e7f475f17fda32b/server/etcdserver/server.go#L1745). (see the [discussion](#push-vs-poll-when-leader-decides-cluster-feature-gate) about why we choose to push the information through raft)
+
+1. When the leader received member attribute update from raft, it will decide the values for the cluster level features based on the proposed feature values of all members.
+
+1. The leader sends the final feature values in `ClusterFeatureGateSetRequest` through raft. 
+
+1. Each member apply the updates to their cluster feature gate, and save the results in the `cluster` bucket in the backend.
+
+![cluster feature gate consensus algorithm](./cluster_feature_gate.png "cluster feature gate consensus algorithm")
```

Comment:
I see you added a new step to graphs that is not mentioned in the algorithm. That on cluster changes we reset ClusterFeatureGate to nil. Does it mean that when we do any cluster changes (resize), we disable all features? Wouldn't that cause unnecessary downtime for critical features and user frustration?

Reply:
Outdated.